aten/src/ATen/native/TensorIndexing.h

```diff
@@ -77,36 +125,454 @@ struct CAFFE2_API TensorIndex final {
+// is undesirable.
+static inline int64_t count_specified_dimensions(const ArrayRef<TensorIndex>& indices) {
+  // Count the number of indexed dimensions (everything but ellipsis and None)
+  int64_t count = 0;
+  size_t size = indices.size();
+  for (auto& obj : indices) {
+    if (obj.is_tensor()) {
+      auto& tensor = obj.tensor();
+      if (tensor.scalar_type() == kByte || tensor.scalar_type() == kBool) {
+        count += tensor.dim();
+      } else {
+        count++;
+      }
+    } else if (!obj.is_none() && !obj.is_ellipsis() && !obj.is_boolean()) {
+      count++;
+    }
+  }
+  return count;
+}
+} // namespace impl
+
+// NOTE: Many functions below are only for consumption from Python indexing
+// implementation, they include:
+//
+// - `Tensor scalarToTensor(...)`
+// - `IntArrayRef slicePrefix1sSize(...)`
+// - `void copy_to(...)`
+// - `Tensor handleDimInMultiDimIndexing(...)`
+// - `Tensor dispatch_index(...)`
+// - `Tensor dispatch_index_put_(...)`
+// - `Tensor get_item(...)`
+// - `void set_item(...)`
+//
+// The rest of the functions are in `at::indexing::impl` namespace, signifying
+// that they shouldn't be used from Python indexing implementation.
+static inline Tensor scalarToTensor(Scalar v, const TensorOptions& options, const at::Device& self_device) {
+  if (self_device == at::kCPU || self_device == at::kCUDA) {
+    return impl::scalarToTensorCPUOrCUDA(v, options);
+  } else {
+    return impl::scalarToTensorNonNativeDeviceType(v, options);
+  }
+}
+
+// To match numpy semantics:
+// As a special case for backwards compatibility,
+// strip away unit dimensions from the left of 'src'
+static inline IntArrayRef slicePrefix1sSize(const IntArrayRef& sizes) {
+  size_t first_non1_src = sizes.size();
+  for (size_t i = 0; i < sizes.size(); ++i) {
+    if (sizes[i] != 1) {
+      first_non1_src = i;
+      break;
+    }
+  }
+
+  return sizes.slice(first_non1_src);
+}
+
+static inline void copy_to(const Tensor& dst, const Tensor& src) {
+  Tensor b_src;
+  std::tie(b_src) = expand_inplace(dst, src.view(slicePrefix1sSize(src.sizes())), "setitem");
+  dst.copy_(b_src);
+}
+
+static inline Tensor handleDimInMultiDimIndexing(
+    const Tensor& prev_dim_result,
+    const Tensor& original_tensor,
+    const TensorIndex& index,
+    int64_t* dim_ptr,
+    int64_t* specified_dims_ptr,
+    int64_t real_dim,
+    std::vector<Tensor>& outIndices,
+    bool is_tracing,
+    const at::Device& original_tensor_device,
+    const IntArrayRef& prev_dim_result_sizes) {
+  if (index.is_integer()) {
+    return impl::applySelect(prev_dim_result, *dim_ptr, index.integer(), real_dim, original_tensor_device, prev_dim_result_sizes);
+  } else if (index.is_slice()) {
+    Tensor result = impl::applySlice(
+      prev_dim_result,
+      *dim_ptr,
+      index.slice().start(),
+      index.slice().stop(),
+      index.slice().step(),
+      /*ensure_view=*/false,
+      /*is_tracing=*/is_tracing,
```

Comment:
On the other hand, this `is_tracing` matters! (Because it is *setters* that attempt to apply the optimization)

Reply:
`/* ensure_view */is_tracing`